Feature: Number sample predict

garland-culbreth opened the pull request:
# Changes

Add a new `number_sample_predict` option to option_predict.csv which specifies the number of samples used by the `csv.predict` command. If not specified, `number_sample` from option_fit.csv is used to preserve backwards compatibility.

Add documentation for the new `number_sample_predict` option in the XRST block at the beginning of at_cascade/csv/predict.py.

# Tests

Add a new test:

```txt
/test
└─ /csv
   └─ number_sample_predict.py
```

The test specifies option_predict.csv's `number_sample_predict` greater than option_fit.csv's `number_sample`, runs `csv.fit` and `csv.predict`, and asserts that every row in the sam_predict.csv produced by csv.predict's output table has `sample_index < number_sample_predict`.

The new test passes. An example of its stdout is attached as: 
[20250630T160735.log](https://github.com/user-attachments/files/20988252/20250630T161535.log)

With this new test added, I ran bin/check_all.sh and it passed fully. The stdout of that test run is attached as: 
[20250630T162318.log](https://github.com/user-attachments/files/20988289/20250630T162318.log)

garland-culbreth commented:
This is marked as a draft because I'm unable to compile the xrst documentation. Running bin/run_xrst.sh and executing `xrst` directly both give me a locale error:

```
(venv) root@4c62e51e20f0:/home/at_cascade.git# bin/run_xrst.sh 
python -m xrst --local_toc --html_theme sphinx_rtd_theme --index_page_name at_cascade --group_list default --number_jobs 8 --target html

sphinx-build -b html -j 1 build/rst build/html

Traceback (most recent call last):
  File "/home/venv/bin/sphinx-build", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/venv/lib/python3.12/site-packages/sphinx/cmd/build.py", line 476, in main
    locale.setlocale(locale.LC_ALL, '')
  File "/usr/lib/python3.12/locale.py", line 615, in setlocale
    return _setlocale(category, locale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
locale.Error: unsupported locale setting
Error: see messages above.
```

This was run in my docker container.

bradbell commented:
I just tried the following and it worked for me:

podman run dismod_at.at_cascade.release -it
pip install tomli
pip install xrst
pip uninstall -y xrst
pip install --index-url https://test.pypi.org/simple/ xrst
pip install sphinx_rtd_theme
cd ../at_cascade.git
bin/run_xrst.sh

I am not sure why pip was not picking up the tomli library as a dependency for xrst (in this context).

Check your xrst,  dismod_at, and  at_cascade and  versions mine are and xrst-2025.6.12, dismod_at-2025.6.17, at-cascade-2025.6.3 
xrst --version
dismod_at

python
import at_cascade
print( at_cascade.version)
quit()

garland-culbreth commented:
Re-ran after setting up sphinx in my image tree and now sphinx and xrst both work properly. My steps were as above:

```txt
pip install tomli
pip install xrst
pip uninstall -y xrst
pip install --index-url https://test.pypi.org/simple/ xrst
pip install sphinx_rtd_theme
```

and then bun/run_xrst.sh ran successfully:

```txt
root@c281b0a12c17:/home/at_cascade# bin/run_xrst.sh
python -m xrst --local_toc --html_theme sphinx_rtd_theme --index_page_name at_cascade --group_list default --number_jobs 8 --target html
Using following input_files: git ls-files
sphinx-build -b html -j 8 build/rst build/html
rm -r build/html/_sources
cp -r build/rst/_sources build/html/_sources
xrst: OK
run_xrst.sh: OK
```

It seems like the issue is something to do with opening the container in VScode vs my terminal: When I use VScode's terminal I get the locale error still, but when I use my terminal I get no error.

bradbell commented:
Looking at number_sample_predict.py there is miss-match in the definition of young. At age 20 in covariate.csv young is true, but it is false at age 20 in the code
```
      young          = float( age < 20 )
```
If you are going to have code in two places that determine the value of young, you should have a comment saying that the values must agree.

garland-culbreth commented:
I opted to removed the `young` covariate from this test altogether, since the covariate isn't required for this test case and removing it simplifies the test further.

I'll try to find some time to make a new branch and look for other occurrences of this and address those too.

bradbell commented:
In number_sample_predict.py, the following comment appears:
```
# Use node n-1 to test starting below the base of the tree
```
There is not n-1 in this example. I suggest that you read through number_sample_predict.py. The comments should mention which variables (that are set below) are  going to be used later (ones not mentioned are just temporaries). This helps me understand a program when I come back to it years later.

garland-culbreth commented:
The two commits above now have the new test passing, but `bin/check_all.sh` fails during `example/csv/binomial.py` due to:

```txt
File "/home/at_cascade_gc.git/at_cascade/csv/pre_one_process.py", line 347, in pre_one_process
    number_sample_predict = int(option_predict['number_sample_predict'])
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'Sincidence'
```

So it seems I'm not assigning `number_sample_predict` properly everywhere it needs to be and am tracking that down.

garland-culbreth commented:
Fixed the issue mentioned in the last comment. Now `bin/check_all.sh` passes. Example: 
[20250715T165215_check_all.log](https://github.com/user-attachments/files/21245276/20250715T165215_check_all.log)

I also streamlined `test/csv/number_sample_predict.py` and cleaned up the comments within it.

garland-culbreth commented:
In https://github.com/bradbell/at_cascade/pull/33/commits/09b1098b899caafa5bddf6f795094cff4d263939 I added the log table check determining which fit type passed just directly before the prediction. It'd be very simple to abstract that into a function which might be useful elsewhere for similar checks. Would it be worth making a new module to house that function? I guess it could go under `./at_cascade`, e.g., something like `at_cascade.match_string_in_log_table`, absent a dedicated utils submodule.

garland-culbreth commented on at_cascade/csv/pre_one_job.py:
```diff
@@ -85,6 +85,14 @@
+This integer option specifies the number of samples generated for each
+prediction. This value is specified in
+:ref:`csv.predict@Input Files@option_predict.csv@number_sample_predict`. The
+default value is :ref:`csv.fit@Input Files@option_fit.csv@number_sample` from
```
Meant to go back and fix that but forgot earlier, thanks for catching. Fixed by https://github.com/bradbell/at_cascade/pull/33/commits/626d1e77087ae691d9f82bac2f4a90f0e9f7f63d now.

JosephWagner commented on at_cascade/csv/pre_one_job.py:
```diff
@@ -398,6 +411,34 @@ def pre_one_job(
+   try:
+      command = [
+         'dismod_at',
+         pre_database,
```
i'm not familiar with the internals of Dismod-AT sqlite databases... does sampling write to pre_database in such a way that we need to worry about overwriting prior sample results?

I suppose if pre_database is a copy that exists only for this prediction operation, that should be fine?

JosephWagner replied:
Oh now that I wrote that out, I suppose mutating the database is the entire point of this? It looks like we rely on samples from pre_database later?

garland-culbreth replied:
Yeah, the predict command later in takes the sample table as an input and operates on it. There's a few calls to the sample command during fit_one_job.py, and the samples generated from those get moved around for other uses, but here we wanted to re-populate the sample table so that when the predict command operates on it it'll have the new number of samples, instead of the number of samples produced when fit_one_job ran the sample command.

JosephWagner commented on at_cascade/csv/pre_one_job.py:
```diff
@@ -398,6 +411,34 @@ def pre_one_job(
+      ]
+      dismod_at.system_command_prc(command)
+   except Exception as e:
+      print(f'sample {fit_type} message: {str(e)}')
```
why do we not raise here but we do raise earlier? If there's some invariant we need to preserve around not raising, would those earlier ValueErrors be an issue?

garland-culbreth replied:
I'm actually not sure why either, I'd initiall lifted this structure from the sample command elsewhere but now I can't find or recall where. I switched it to a proper raise to be more internally consistent. 

This also led me to catch an oversight with the earlier raises, if they trigger the database connection doesn't get closed.

These changes in https://github.com/bradbell/at_cascade/pull/33/commits/897fab8ffad17471aaa2f8c3adbf9ad5f78d8e07.

JosephWagner commented:
looks good to me. I had a few small comments around docs and some error handling, but nothing major on my end (as an outside reviewer)

garland-culbreth commented:
Re-ran test/csv/number_sample_predict.py and bin/check_all.sh with these changes. Both are passing.

Log output from bin/check_all.sh: [20250730T125703-check-all.log](https://github.com/user-attachments/files/21517091/20250730T125703-check-all.log)

bradbell commented:
It seems possible to me, in an arbitrary dismod_at log, that a fit could succeed and a later fit fail. I therefore find the following (in pre_one_job) confusing:
```
  log_table = dismod_at.get_table_dict(connection, 'log')
   log_fit_types = []
   fit_passed = False
   for row in log_table:
      if 'begin fit' in row['message']:
         log_fit_types.append(row['message'].replace('begin fit', '').strip())
      if 'end fit' in row['message']:
         fit_passed = True
```
I think that this would be simplified by may added the command arguments to the `end fit` log messages (and other end command log messages). Then one could just check for the last 'end fit' and make sure the was no `begin fit` after it.
